api/filter/fault.proto

```diff
@@ -0,0 +1,127 @@
+    FIXED = 0;
+    // Exponential delay.
+    EXPONENTIAL = 1;
+  }
+
+  // Delay type to use (fixed|exponential|..). Currently, only fixed delay (step function) is supported.
+  FaultDelayType type = 1;
+
+  // An integer between 0-100 indicating the percentage of operations/connection requests
+  // on which the delay will be injected.
+  google.protobuf.UInt32Value percent = 2;
+
+  oneof fault_delay_type {
+    // Add a fixed delay before forwarding the operation upstream. See
+    // https://developers.google.com/protocol-buffers/docs/proto3#json for
+    // the JSON/YAML Duration mapping. For HTTP/Mongo/Redis, the specified
+    // delay will be injected before a new request/operation. For TCP
+    // connections, the proxying of the connection upstream will be delayed
+    // for the specified period.
+    google.protobuf.Duration fixed_delay = 3;
+  }
+}
+
+// Abort specification is used to prematurely abort a request/TCP connection
+// with a pre-specified error code.
+message FaultAbort {
+  // An integer between 0-100 indicating the percentage of requests/operations/connections
+  // that will be aborted with the error code provided.
+  google.protobuf.UInt32Value percent = 1;
+
+  // Applicable only for HTTP connections.
+  oneof error_type {
+    // HTTP status code to use to abort the HTTP request.
+    int32 http_status = 2;
+    // gRPC status code to use to abort the gRPC request.
+    int32 grpc_status = 3;
+  }
+}
+
+// The fault injection filter can be used to test the resiliency of
+// microservices to different forms of failures. The filter can be used to
+// inject delays and abort requests with user-specified error codes,
+// thereby providing the ability to stage different failure scenarios such
+// as service failures, service overloads, high network latency, network
+// partitions, etc. Faults injection can be limited to a specific set of
+// requests based on the (destination) upstream cluster of a request and/or
+// a set of pre-defined request headers.
+//
+// The scope of failures is restricted to those that are observable by an
+// application communicating over the network. CPU and disk failures on the
+// local host cannot be emulated.
+//
+// Currently, the fault injection filter has the following limitations:
+//
+// * Abort codes are restricted to HTTP status codes only
+// * Delays are restricted to fixed duration.
+//
+// Future versions will include support for restricting faults to specific
+// routes, and delay durations based on distributions.
+//
+// * Note:* The fault injection filter must be inserted before any
+// other filter, including the router filter.
+message HTTPFault {
+  // If specified, the filter will inject delays based on the values in the
+  // object. At least abort or delay must be specified.
+  Delay delay = 1;
+
+  // If specified, the filter will abort requests based on the values in
+  // the object. At least abort or delay must be specified.
+  Abort abort = 2;
```

Comment:
Is there ever a situation in which this or delay would be a `repeated`?

Reply:
Good question. My answer would be no. The filters can be stacked (i.e. multiple fault filters), such that the matching filter will activate. Changing this into an array would require changes to the filter implementation to process a sequence of faults rather than assume that each filter is responsible for one delay + abort.